editors/sc-ide/core/main.cpp

```diff
@@ -135,24 +136,39 @@ int main( int argc, char *argv[] )
     if (startInterpreter)
         main->scProcess()->startLanguage();
 
-    // setup HelpBrowser server
+    // set up HelpBrowser server
+    HelpBrowser* browser = win->helpBrowserDocklet()->browser();
     QWebSocketServer server("SCIDE HelpBrowser Server", QWebSocketServer::NonSecureMode);
-    if (!server.listen(QHostAddress::LocalHost, 12344)) {
-        qFatal("Failed to open web socket server.");
-        return 1;
+    findOpenPort(server);
+    if (server.serverPort()) {
+        browser->setServerPort(server.serverPort());
+
+        // setup comm channel
+        WebSocketClientWrapper clientWrapper(&server);
+        QWebChannel channel;
+        QObject::connect(&clientWrapper, &WebSocketClientWrapper::clientConnected,
+                         &channel, &QWebChannel::connectTo);
+
+        // publish IDE interface
+        IDEWebChannelWrapper ideWrapper{browser};
+        channel.registerObject("IDE", &ideWrapper);
```

Comment:
ehh i'm not on board with that, looks pretty weird. how about setting them as members of `HelpBrowser` instead?

Reply:
making them members means you now need to manage extra state (i.e. are these objects default-constructed and therefore invalid, or valid; maybe even need to write those default constructors an and state accessors). this goes against RAII and is not idiomatic C++. it's not worth it when you could be much clearer by adding one line here. besides, logically these objects do not really belong to the help browser if it can be constructed without them. there could be some sort of RAII wrapper that encapsulates this logic but I would rather be explicit about that than put it in a logically distinct object.

Reply:
well alright -- but for the record, i really dislike the two copies of `return app.exec()`!

Reply:
revisiting this -- one way to clean this a bit would be by creating a small wrapper class with pointers for services that are only valid if the server listens. the design is fairly straightforward IMO and encapsulates this logic well. dunno what you think.

```cpp
// in main.cpp
class HelpBrowserWebSocketServices {

public:
    HelpBrowserWebSocketServices(HelpBrowser* browser) :
    {
        m_server = new QWebSocketServer("SCIDE HelpBrowser Server", QWebSocketServer::NonSecureMode);
        if (m_server->listen(QHostAddress::LocalHost)) {
            browser->setServerPort(m_server->serverPort());

            // setup comm channel
            m_clientWrapper = new WebSocketClientWrapper(m_server);
            m_channel = new QWebChannel();
            QObject::connect(m_clientWrapper, &WebSocketClientWrapper::clientConnected, m_channel, &QWebChannel::connectTo);

            // publish IDE interface
            m_ideWrapper = new IDEWebChannelWrapper(browser);
            m_channel.registerObject("IDE", m_ideWrapper);
        } else {
            qWarning("Failed to set up help browser web socket.");
            // [don't take up heap space with a useless object]
            delete m_server;
            m_server = nullptr;
        }
    }

    ~HelpBrowserWebSocketServices()
    {
        // [delete on nullptr is valid and no-op]
        delete m_server;
        delete m_clientWrapper;
        delete m_channel;
        delete m_ideWrapper;
    }

private:
    // non-null only if server listens successfully
    QWebSocketServer* m_server;
    WebSocketClientWrapper* m_clientWrapper;
    QWebChannel* m_channel;
    IDEWebChannelWrapper* m_ideWrapper;
};
```

then this part becomes:

```cpp
    // line 130
    if (startInterpreter)
        main->scProcess()->startLanguage();

    HelpBrowserWebSocketServices services(win->helpBrowserDocklet()->browser());
    return app.exec();
```

Reply:
this would also make things cleaner in #4328